pkg/blobstore/quorum/quorum_blob_access.go

```diff
@@ -0,0 +1,218 @@
+        return eh.tryNextBackendOrError(fallbackErr)
+    }
+
+	return nil, err
+}
+
+func (eh getQuorumErrorHandler) Done() {}
+
+func (ba *quorumBlobAccess) get(getter func(b blobstore.BlobAccess) buffer.Buffer) buffer.Buffer {
+	backends := ba.shuffledBackends()
+
+	backend := backends[0]
+	remainingBackends := backends[1:]
+
+	return buffer.WithErrorHandler(
+		getter(backend),
+		&getQuorumErrorHandler{
+			remainingBackends: remainingBackends,
+			remainingReads:    ba.readQuorum,
+			retry:             getter,
+		})
+}
+
+func (ba *quorumBlobAccess) Get(ctx context.Context, digest digest.Digest) buffer.Buffer {
+	return ba.get(func(b blobstore.BlobAccess) buffer.Buffer {
+		return b.Get(ctx, digest)
+	})
+}
+
+func (ba *quorumBlobAccess) GetFromComposite(ctx context.Context, parentDigest, childDigest digest.Digest, slicer slicing.BlobSlicer) buffer.Buffer {
+	return ba.get(func(b blobstore.BlobAccess) buffer.Buffer {
+		return b.GetFromComposite(ctx, parentDigest, childDigest, slicer)
+	})
+}
+
+func (ba *quorumBlobAccess) shuffledBackendQueue() <-chan blobstore.BlobAccess {
+	queue := make(chan blobstore.BlobAccess)
+	go func() error {
+		backends := ba.shuffledBackends()
+
+		for _, b := range backends {
+			queue <- b
+		}
+
+		close(queue)
+		return nil
+	}()
+    return queue
+}
+
+func (ba *quorumBlobAccess) Put(ctx context.Context, digest digest.Digest, b buffer.Buffer) error {
+	// Store object in at least writeQuorum storage backends.
+	group, ctx := errgroup.WithContext(ctx)
+    backendQueue := ba.shuffledBackendQueue()
+
+	// Spawn writeQuorum writers.  Each of these goroutines needs to succeed once.
+	for i := 0; i < ba.writeQuorum; i++ {
+		var b1 buffer.Buffer
+		if i == ba.writeQuorum-1 {
+			// Last writer, no need to clone buffer
+			b1 = b
+		} else {
+			b, b1 = b.CloneStream()
+		}
+
+		group.Go(func() error {
+			var err error
+			for backend := range backendQueue {
+				err = backend.Put(ctx, digest, b1)
+				if err == nil {
+					// Success
+					return nil
+				}
+			}
+			return err
+		})
+	}
+
+	return group.Wait()
+}
+
+func (ba *quorumBlobAccess) FindMissing(ctx context.Context, digests digest.Set) (digest.Set, error) {
+	// Call FindMissing() on readQuorum backends.
+	group, ctx := errgroup.WithContext(ctx)
+    backendQueue := ba.shuffledBackendQueue()
+
+	results := make([]digest.Set, ba.readQuorum)
+	for i := 0; i < ba.readQuorum; i++ {
+		resultIdx := i
+		group.Go(func() error {
+			var err error
+			for backend := range backendQueue {
+				results[resultIdx], err = backend.FindMissing(ctx, digests)
+				if err == nil {
+					// Success
+					return nil
+				}
+			}
+			return err
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		return digest.EmptySet, err
+	}
+
+	// Find intersection of all results
+	missingFromAll := results[0]
+	for _, result := range results[1:] {
+		_, missingFromAll, _ = digest.GetDifferenceAndIntersection(missingFromAll, result)
+	}
+	return missingFromAll, nil
```

Comment:
Imagine I have 5 storage backends, named 0 to 4. Read quorum is set to 4.

- I first issue a Put(). Due to network flakiness, the blob only ends up getting written to backend 0.-
- Then I issue FindMissingBlobs(), which gets run against backends 0, 1, 2, and 3. Backend 0 reports it as present, while backends 1, 2 and 3 report it as absent. As we compute the intersection, we report it as being present.
- Then I issue Get(), which gets run against backends 1, 2, 3 and 4. All of those return NOT_FOUND.

This is in violation of the protocol.